src/reseeding.rs

```diff
@@ -51,43 +55,85 @@ impl<R: Rng, Rsdr: Reseeder<R>> ReseedingRng<R, Rsdr> {
 
     /// Reseed the internal RNG if the number of bytes that have been
     /// generated exceed the threshold.
-    pub fn reseed_if_necessary(&mut self) {
+    pub fn reseed_if_necessary(&mut self) -> Result<(), Error> {
         if self.bytes_generated >= self.generation_threshold {
-            self.reseeder.reseed(&mut self.rng);
+            let _ = self.reseeder.reseed(&mut self.rng).map_err(map_err)?;
+            // Reset number of generated bytes, even if reseeding failes.
+            // This means we don't try to reseed again until a new
+            // `generation_threshold` amount of output is consumed.
             self.bytes_generated = 0;
-        }
+        };
+        Ok(())
+    }
+}
+
+/// An error that can occur when reseeding failed.
+#[cfg(feature="std")]
+#[derive(Debug)]
+pub struct ReseedingError {
+    cause: Box<::std::error::Error + Send + Sync>,
+}
+
+#[cfg(feature="std")]
+impl fmt::Display for ReseedingError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Failed to reseed RNG.")
     }
 }
 
+#[cfg(feature="std")]
+impl ::std::error::Error for ReseedingError {
+    fn description(&self) -> &str {
+        "reseeding failed"
+    }
+
+    fn cause(&self) -> Option<&::std::error::Error> {
+        Some(self.cause.as_ref())
+    }
+}
+
+#[cfg(feature="std")]
+fn map_err<E>(_cause: E) -> Error
+    where E: Into<Box<::std::error::Error + Send + Sync>>
+{
+    let cause = ReseedingError { cause: _cause.into() };
+    Error::new_err(ErrorKind::Other, cause)
+}
+
+#[cfg(not(feature="std"))]
+fn map_err<E>(_cause: E) -> Error {
+    // We can't pass on the cause in `no_std` mode
+    Error::new_str(ErrorKind::Other, "reseeding failed")
+}
 
 impl<R: Rng, Rsdr: Reseeder<R>> Rng for ReseedingRng<R, Rsdr> {
     fn next_u32(&mut self) -> u32 {
-        self.reseed_if_necessary();
+        let _ = self.reseed_if_necessary(); // continue if reseeding failed
         self.bytes_generated += 4;
         self.rng.next_u32()
     }
 
     fn next_u64(&mut self) -> u64 {
-        self.reseed_if_necessary();
+        let _ = self.reseed_if_necessary(); // continue if reseeding failed
         self.bytes_generated += 8;
         self.rng.next_u64()
     }
 
     #[cfg(feature = "i128_support")]
     fn next_u128(&mut self) -> u128 {
-        self.reseed_if_necessary();
+        let _ = self.reseed_if_necessary(); // continue if reseeding failed
         self.bytes_generated += 16;
         self.rng.next_u128()
     }
 
     fn fill_bytes(&mut self, dest: &mut [u8]) {
-        self.reseed_if_necessary();
+        let _ = self.reseed_if_necessary(); // continue if reseeding failed
         self.bytes_generated += dest.len() as u64;
         self.rng.fill_bytes(dest);
     }
 
     fn try_fill(&mut self, dest: &mut [u8]) -> Result<(), Error> {
-        self.reseed_if_necessary();
+        self.reseed_if_necessary()?;
```

Comment:
I'm not keen on this difference of error handling between functions. Either the `ReseedingRng` should always allow continuation when reseeding fails (in which case this function should either silently continue or report `ErrorKind::Transient` and have some extra logic not to try reseeding too soon after a failure), or `ReseedingRng` should always fail if reseeding fails (with panic if necessary).

I'm not sure which; it could even be configurable but probably unnecessary. Perhaps allowing continuation after failure is the best option (users shouldn't rely on reseeding for security in any case).

Reply:
Good point, we can return an error here _and_ fill `dest`. Then it is up to the calling code to decide what to do.